corehq/apps/app_manager/app_translations/validator.py

```diff
@@ -0,0 +1,86 @@
+from __future__ import absolute_import
+
+from memoized import memoized
+from corehq.apps.app_manager.app_translations import (
+    expected_bulk_app_sheet_headers,
+    expected_bulk_app_sheet_rows,
+    get_unicode_dicts,
+)
+from corehq.apps.app_manager.app_translations.const import MODULES_AND_FORMS_SHEET_NAME
+
+COLUMNS_TO_COMPARE = {
+    'module_and_form': ['Type', 'sheet_name'],
+    'module': ['case_property', 'list_or_detail'],
+    'form': ['label'],
+}
+ROW_COUNT_MISMATCH_MESSAGE = "We found {} rows than expected. Validating what was uploaded."
+
+
+class UploadedTranslationsValidator(object):
+    """
+    this compares the excel sheet uploaded with translations with what would be generated
+    with current app state and flags any discrepancies found between the two
+    """
+    def __init__(self, app, uploaded_workbook, lang_prefix='default_'):
+        self.app = app
+        self.uploaded_workbook = uploaded_workbook
+        self.headers = None
+        self.expected_rows = None
+        self.lang_prefix = lang_prefix
+        self.default_language_column = self.lang_prefix + self.app.default_language
+
+    def _generate_expected_headers_and_rows(self):
+        self.headers = {h[0]: h[1] for h in expected_bulk_app_sheet_headers(self.app)}
+        self.expected_rows = expected_bulk_app_sheet_rows(self.app)
+
+    @memoized
+    def _get_header_index(self, sheet_name, header):
+        for index, _column_name in enumerate(self.headers[sheet_name]):
+            if _column_name == header:
+                return index
+
+    def _compare_sheet(self, sheet_name, uploaded_rows, for_type):
+        """
+        :param uploaded_rows: dict
+        :param for_type: type of sheet, module_and_forms, module, form
+        :return: list of errors messages if any
+        """
+        columns_to_compare = COLUMNS_TO_COMPARE[for_type] + [self.default_language_column]
+        expected_rows = self.expected_rows[sheet_name]
+        msg = []
```

Comment:
nit: I'd expect `msg` to refer to a single message, rather than a list of messages.

Reply:
replaced with `msgs`